### PR TITLE
Add ttl parameter to UDP and ICMP write

### PR DESCRIPTION
**Pull request by phaer**
I'd like to expose the optional `ttl` parameter of the underlying IP.write to ICMP.write and UDP.write, because I think its helpful to implement traceroute functionality.

* `ICMP.write`: add ttl parameter.

Talked about it for a moment with @hannesm at the retreat.

### CODE REVIEW

File: src/mirage_protocols.mli [hunk 1]
```diff
@@ -236,7 +236,7 @@ module type ICMP = sig
   (** [input t src dst buffer] reacts to the ICMP message in
       [buffer]. *)
 
-  val write : t -> dst:ipaddr -> buffer -> (unit, error) result io
+  val write : t -> dst:ipaddr -> ?ttl:int -> buffer -> (unit, error) result io
   (** [write t dst buffer] sends the ICMP message in [buffer] to [dst]
```
Comment (hannesm) [inline src/mirage_protocols.mli]: the documentation string should be updated, please

File: src/mirage_protocols.mli [hunk 2]
```diff
@@ -283,7 +283,7 @@ module type UDP = sig
       return a concrete handler or a [None], which results in the
       datagram being dropped. *)
 
-  val write: ?src_port:int -> dst:ipaddr -> dst_port:int -> t -> buffer ->
+  val write: ?src_port:int -> ?ttl:int -> dst:ipaddr -> dst_port:int -> t -> buffer ->
```
Comment (hannesm) [inline src/mirage_protocols.mli:286]: see above

File: src/mirage_protocols.mli (hunk 1 above)
Comment (phaer) [reply src/mirage_protocols.mli]: Done, thanks for catching that!

File: src/mirage_protocols.mli (hunk 2 above)
Comment (phaer) [reply src/mirage_protocols.mli:286]: Done, thanks

File: src/mirage_protocols.mli [hunk 3]
```diff
@@ -236,9 +236,9 @@ module type ICMP = sig
   (** [input t src dst buffer] reacts to the ICMP message in
       [buffer]. *)
 
-  val write : t -> dst:ipaddr -> buffer -> (unit, error) result io
-  (** [write t dst buffer] sends the ICMP message in [buffer] to [dst]
-      over IP. *)
+  val write : t -> dst:ipaddr -> ?ttl:int -> buffer -> (unit, error) result io
+  (** [write t dst ?ttl buffer] sends the ICMP message in [buffer] to [dst]
```
Comment (hannesm) [inline src/mirage_protocols.mli]:
```suggestion
  (** [write t dst ~ttl buffer] sends the ICMP message in [buffer] to [dst]
```
the `[ ... ]` is a call-site of a function, where usually `~` is used for labeled arguments.

File: src/mirage_protocols.mli [hunk 4]
```diff
@@ -236,9 +236,9 @@ module type ICMP = sig
...
+      over IP. Passes [ttl] to the IP stack if given. *)
```
Comment (hannesm) [inline src/mirage_protocols.mli]:
```suggestion
      over IP. Passes the time-to-live ([ttl]) to the IP stack if given. *)
```

File: src/mirage_protocols.mli [hunk 5]
```diff
@@ -283,11 +283,12 @@ module type UDP = sig
       return a concrete handler or a [None], which results in the
       datagram being dropped. *)
 
-  val write: ?src_port:int -> dst:ipaddr -> dst_port:int -> t -> buffer ->
+  val write: ?src_port:int -> ?ttl:int -> dst:ipaddr -> dst_port:int -> t -> buffer ->
     (unit, error) result io
-  (** [write ~src_port ~dst ~dst_port udp data] is a thread
+  (** [write ~src_port ?ttl ~dst ~dst_port udp data] is a thread
```
Comment (hannesm) [inline src/mirage_protocols.mli]:
```suggestion
  (** [write ~src_port ~ttl ~dst ~dst_port udp data] is a thread
```

File: src/mirage_protocols.mli [hunk 6]
```diff
@@ -283,11 +283,12 @@ module type UDP = sig
...
       that writes [data] from an optional [src_port] to a [dst]
-      and [dst_port] IPv4 address pair. *)
+      and [dst_port] IPv4 address pair. An optional [ttl] is passed
```
Comment (hannesm) [inline src/mirage_protocols.mli]:
```suggestion
      and [dst_port] IPv4 address pair. An optional time-to-live ([ttl]) is passed
```

**Review by hannesm — COMMENTED** [review]
thanks, minor suggestions above!

**Comment by phaer** [comment]
Thanks for the thorough feedback. All suggestions applied.

**Comment by hannesm** [comment]
thanks, squashed and merged as part of #20

**Pull request closed without merging**